### ISSUE TITLE
Add a Parameter node label: 12 languages already extract parameters, the data is discarded

### ISSUE THREAD

**Issue by vitali87** [vitali87/code-graph-rag#1804]
## Summary

Add a `Parameter` node label. Twelve languages already have complete, edge-case-hardened parameter extractors in the repo; they are gated off by default and their output is discarded. This issue persists what is already computed.

## Why this one first

It is the cheapest of the four granularity issues and needs no new per-language parsing. The extractors already handle: varargs, Go grouped declarations (`a, b int`), TypeScript `this` parameters occupying no slot, Rust `self_parameter` exclusion, C `variadic_parameter`, Dart `optional_formal_parameters` flattening, PHP `$`-prefixed names, Java `spread_parameter` varargs with the name in a nested declarator.

## What exists today

**Parameter names — 12 languages**, `*_positional_parameter_slots` in `codebase_rag/parsers/utils.py`:

| Language | utils.py |
|---|---|
| Go | 807 |
| JS/TS | 871 |
| C++ | 895 |
| Java | 914 |
| C# | 951 |
| Lua | 1000 |
| Scala | 1024 |
| Rust | 1053 |
| C | 1074 |
| PHP | 1104 |
| Dart | 1129 |
| Python | `flow_access/processor.py:142` |

Dispatched by `_lean_parameter_slots`, `codebase_rag/parsers/flow_access/processor.py:171-204`. Each returns `tuple[list[str | None], int | None]` — one entry per formal slot, `None` where the slot binds no simple name, plus the variadic index.

**Parameter types — 9 languages**, already persisted as the parallel `param_types: list[string]?` property on `Function`/`Method` (`types_defs.py:895`, `:899`), extracted in `codebase_rag/parsers/type_facts.py:192-203`. Per `type_facts.py:15-18` the array is parallel to declared parameters in source order, `""` for unannotated.

**Type resolution** already works: `TypeReferenceResolver` (`type_facts.py:264-357`) resolves an annotation string to a project type QN via imports, module scope, then unique-suffix match.

## Blockers

1. **Gated off by default.** `flow_access/processor.py:633` enables the slot extractors only when `FLOWS_TO` is enabled; `CaptureGroup.IO` is not in `DEFAULT_CAPTURE_GROUPS` (`constants/graph.py:366-373`).
2. **No positions retained.** Every extractor returns bare `str`/`str | None`. No `start_point` is read anywhere. This is the main new work: ~1 line per extractor.
3. **`_emit_accepts` deduplicates.** `type_facts.py:388-394` collapses `param_types` into a deduped set and emits one `ACCEPTS` edge per distinct type, deliberately discarding per-parameter identity.
4. **Two languages missing types**: C/C++ return `TypeFacts(ret, None)` (`type_facts.py:186-189`); PHP, Lua, Dart, Scala, SQL are absent from `_EXTRACTORS`.

## Proposed shape

```
Parameter {qualified_name, name, index, path, absolute_path,
           start_line?, start_col?, type_name?, is_variadic?, has_default?}

Function|Method -[:HAS_PARAMETER {index}]-> Parameter
Parameter       -[:OF_TYPE]-> Class|Interface|Type|Enum|Union
```

`self`/`cls` should NOT become nodes — measured 6,050 of 24,787 parameters in this repo.

## Cost

Measured on this repo with `ast`:

- source only (`codebase_rag/`, excl. tests): ~2,450 Function/Method/Class today → ~7,500 with parameters, **3.1x**
- full repo incl. tests: 15,501 → ~40,500, **2.6x**
- real parameters excluding `self`/`cls`: 18,737; **96% are type-annotated**, so most get a resolvable `OF_TYPE` edge

Edge cost is the sharper concern: `TypeReferenceResolver.resolve` would run per parameter rather than once per deduped annotation set.

## Registration checklist

`NodeLabel` + `_NODE_LABEL_UNIQUE_KEYS` (a module-import `RuntimeError` fires if missing, `constants/graph.py:243-247`), `NODE_SCHEMAS`, `RelationshipSchema`, new `RelationshipType` + its `CAPTURE_GROUP_RELS` entry, `codec/schema.proto` message + `Node` oneof field (next free is 22), and `LABEL_TO_ONEOF_FIELD` in `services/protobuf_service.py:13-35` — **omitting the last one makes `ensure_node_batch` silently drop the node** (`protobuf_service.py:112-114`).

Gates: `test_every_declared_property_is_exported_or_declared_unexported`, `test_every_documented_label_has_parsed_properties`, plus the relationship parity tests.

Suggest putting it behind a capture group so it is opt-in initially.

**Comment by vitali87** [comment]
Taking this. One correction to the registration checklist before anyone
writes the proto, since these numbers are append-only wire format and a
collision is expensive to undo:

**"next free is 22" is now stale.** Field 22 was taken by `Gloss` in #1830
(merged). Verified against `codec/schema.proto` on current `main`:

```
127:      CodeSmell code_smell = 20;
128:      SecurityIssue security_issue = 21;
129:      Gloss gloss = 22;
```

So `Parameter` takes payload field **23**. The relationship enum is at 31
(`ANNOTATES = 30`, `MENTIONS = 31`), so `HAS_PARAMETER` and `OF_TYPE` take
**32** and **33**.

Regenerate with the pinned toolchain, not whatever protoc is on PATH:

```
uv run --with "grpcio-tools==1.76.0" python -m grpc_tools.protoc \
  -I. --python_out=. --pyi_out=. codec/schema.proto
```

The pin is load-bearing: generating with a protoc newer than the pinned
runtime writes a gencode version stamp that makes `ValidateProtobufRuntimeVersion`
reject the module, breaking `import codec` for everyone, not just the author.

Confirming the rest of the checklist from having just done it for #1808:
`_NOT_IN_PROTO` is a deliberately empty frozenset, so there is no
partial-registration increment — a bare `NodeLabel` fires three parity tests,
two import-time `RuntimeError` guards, and the graph-audit documented-labels
test. It is full registration or nothing, which matches the "suggest putting
it behind a capture group" note: the capture group controls whether the node
is EMITTED, and is orthogonal to whether the label is registered.

Agreed on excluding `self`/`cls`, and on the edge cost being the sharper
concern — I'll measure `TypeReferenceResolver.resolve` per-parameter against
the current per-deduped-set cost before wiring `OF_TYPE`, and report the
number rather than assuming it is affordable.

**Comment by vitali87** [comment]
Measured the edge-cost concern, since "the sharper concern" was the one part
of the spec with no number attached. It is smaller than it reads.

Over this repo's own source (`codebase_rag/`, excluding tests), counting the
`resolve_annotation` calls each strategy implies:

```
functions/methods               3,963
parameter slots (all)           9,905
  self/cls (excluded)           1,984  (20%)
  real parameters               7,921
  annotated                     7,868  (99% of real)

resolve_annotation calls implied:
  today  (deduped per func)     6,807
  proposed (per parameter)      7,868
  factor                         1.16x

distinct annotation strings       575
```

**1.16x, not a multiple.** The intuition behind the concern is that
`_emit_accepts` deduplicates, so per-parameter resolution must cost
proportionally more. It does not, because the dedup is per FUNCTION and
annotations rarely repeat *within one signature*: 7,868 annotated parameters
collapse to only 6,807 distinct-per-function, a 13% saving. The dedup that
looks expensive to give up was not buying much.

The real lever is elsewhere: only **575 distinct annotation strings** across
7,868 parameters, and the distribution is very top-heavy —

```
  2,333  str
    964  Node
    400  Path
    351  str | None
    264  ASTNode
```

So a resolver-level memo keyed on `(annotation, module_qn)` takes the whole
thing well below today's cost, per-parameter or not. `TypeReferenceResolver.
resolve` walks imports, then module scope, then a unique-suffix match; for
`str` that walk is repeated 2,333 times to reach the same answer. I would
rather add that memo than preserve the dedup, and I will report the measured
before/after rather than asserting it.

Two other numbers from the same pass, for the node-count estimate in the
issue: 99% of real parameters here are annotated (the issue says 96% — close,
and the difference is that I exclude `self`/`cls` before taking the ratio),
and `self`/`cls` are 20% of all slots, consistent with the 6,050/24,787
measurement quoted.

Proto numbering correction from my earlier comment now independently
confirmed by another session against `main`: `Gloss gloss = 22` is the
highest payload field, `MENTIONS = 31` the highest relationship enum. So
Parameter takes 23, and `HAS_PARAMETER`/`OF_TYPE` take 32/33.

**Comment by vitali87** [comment]
One design constraint found while reading the extractors, which changes "~1
line per extractor" for the Python side and is worth settling before anyone
writes code.

**The existing slot extractors deliberately do not enumerate every declared
parameter**, so they cannot be the source for `Parameter` nodes as-is.

`_py_positional_param_names` (`flow_access/processor.py:142`) stops at the
first `*args` / `**kwargs` / bare `*`:

```python
if child.type not in _PY_POSITIONAL_PARAM_TYPES:
    # `*`, `*args`, `**kwargs` (or anything unexpected): everything after
    # is variadic-absorbed or keyword-only, so positional mapping stops.
    break
```

and drops a leading `self`/`cls` so indices line up with call-site arguments.
Both choices are correct for what it exists to do — mapping `arg:<index>` to a
parameter for taint flow (#1142, #1169) — and both are wrong for a node that
is supposed to represent a declared parameter:

- every keyword-only parameter would be missing, silently
- `index` would be the call-site position, not the declaration position, so
  `HAS_PARAMETER {index}` would disagree with the source for any function
  with a variadic or a dropped `self`

The same applies to the lean languages via `_lean_parameter_slots`: the
docstring is explicit that it returns "one entry per formal POSITIONAL slot".

Both helpers have exactly one production caller each
(`processor.py:780` and `:844`), both inside the taint walk, so nothing else
depends on their current shape — but changing them to enumerate everything
would break the thing they were built for, and the comments say why in detail.

So the shape I intend, unless the owner prefers otherwise: a SEPARATE
declaration-order enumerator per language that returns every formal parameter
with its position, leaving the taint extractors untouched. That makes the
per-language work larger than the issue estimates (a new small function, not
one line), but it keeps two genuinely different questions apart — "which
parameter does argument N bind to" versus "what parameters does this function
declare" — rather than overloading one helper with a flag and having every
future reader guess which semantics they are getting.

It also means `self`/`cls` exclusion becomes a policy of the Parameter
emitter rather than an artefact inherited from the taint helper, which is
where the issue already says it belongs.

**Comment by vitali87** [comment]
Python is in: #1891 merged at 16b93296 — label, edges, opt-in `parameters`
capture group, proto field 23 / enum 32–33, the Python enumerator, emission at
both ingest sites, deferred `OF_TYPE` with a per-(annotation, module) memo, and
the incremental paths (delete walks include `HAS_PARAMETER`; `OF_TYPE` is
rehydrated keyed on the file; scoped reingest discards stale facts by file).

Still open here: the other eleven enumerators (Go, JS/TS, C++, Java, C#, Lua,
Scala, Rust, C, PHP, Dart), each a `declared_parameters` case in
`parsers/parameter_nodes.py` mirroring `python_declared_parameters`, with the
same three things the review made explicit for Python — declaration order
(not the taint slot table), variadics as one flagged slot, and the receiver
classified by the call site rather than by name. `param_types` on the owner
keeps its own convention (receiver included on a method), so `index` stays
offset by one there; each Parameter carries its own `type_name`.

Three review findings worth carrying into the other languages: a reused
updater and a fresh updater through `reingest` behave differently for
rehydration (key on the file); the deferred pass must run when only parameter
facts are queued (guard now tests every queue); and a batch flush writes
nodes before relationships, so emission follows the owner node.

**Comment by vitali87** [comment]
The other eleven enumerators are up as #1965 (Go, JS/TS/TSX, C, C++, Java, C#, Lua, Scala, Rust, PHP, Dart), each a `*_declared_parameters` case in `parsers/parameter_nodes.py` mirroring the Python one, behind the same opt-in `parameters` capture group.

Two conventions settled there that the issue body did not fix:

- **A slot that binds no simple name keeps its position and yields no node** (an unnamed C/C++/Go parameter, a JS/TS destructuring pattern, a Rust `_` or tuple pattern, a Lua `...`). Compacting would have made `HAS_PARAMETER {index}` disagree with the declaration, which is the defect the Python round found in the taint table.
- **The receiver is excluded by structure, never by name.** Rust `&self` / `self: Box<Self>`, TypeScript `this: T` and Java `C this` take no position; Go's receiver is a separate field and Lua's `:` receiver is implicit. A C# extension method's `this C self` IS caller-supplied (`c.E(x)` or `E(c, x)`) and keeps its slot. Only Python needs `has_receiver` from the call site.

Also PHP names drop the `$` sigil, as `Field` nodes do for promoted properties, and C/C++ `type_name` is the type specifier without declarator marks (`const char *b` → `char`), the reading `extract_type_facts` already gives a return type.

A twelve-language fixture through `GraphUpdater` + the eval store proves every language's own ingest path (the JS/TS module, the deferred C++ out-of-class method, the Dart signature wrapper) reaches the emitter, that `OF_TYPE` resolves each language's `Widget`, and that no Parameter is orphaned. The local review round found the same first-named-child mistake in `extract_type_facts` for Java `final String... xs`; filed as #1964.

**Comment by vitali87** [comment]
Done on main, closing.
- #1891 (merged 16b93296) added the `Parameter` label, `HAS_PARAMETER` / `OF_TYPE`, proto field 23 with enum 32–33, the opt-in `parameters` capture group, the Python enumerator and the incremental paths.
- #1965 (merged) added the other eleven enumerators (Go, JS/TS/TSX, C, C++, Java, C#, Lua, Scala, Rust, PHP, Dart) as `*_declared_parameters` cases in `codebase_rag/parsers/parameter_nodes.py`, 26 references on `origin/main`. It also has a twelve-language fixture through `GraphUpdater`.

The one defect found along the way is tracked separately as #1964.
